src/librustc_incremental/persist/fs.rs

```diff
@@ -0,0 +1,862 @@
+
+        if timestamp > best_candidate.0 {
+            best_candidate = (timestamp, Some(session_dir));
+        }
+    }
+
+    best_candidate.1
+}
+
+fn extract_timestamp_from_session_dir(directory_name: &str)
+                                      -> Result<SystemTime, ()> {
+    if !directory_name.starts_with("sess-") {
+        return Err(())
+    }
+
+    let dash_indices: Vec<_> = directory_name.match_indices("-")
+                                             .map(|(idx, _)| idx)
+                                             .collect();
+    if dash_indices.len() < 2 {
+        return Err(())
+    }
+
+    string_to_timestamp(&directory_name[dash_indices[0]+1 .. dash_indices[1]])
+}
+
+fn timestamp_to_string(timestamp: SystemTime) -> String {
+    let duration = timestamp.duration_since(UNIX_EPOCH).unwrap();
+    let nanos = duration.as_secs() * 1_000_000_000 +
+                (duration.subsec_nanos() as u64);
+    format!("{:x}", nanos)
+}
+
+fn string_to_timestamp(s: &str) -> Result<SystemTime, ()> {
+    let nanos_since_unix_epoch = u64::from_str_radix(s, 16);
+
+    if nanos_since_unix_epoch.is_err() {
+        return Err(())
+    }
+
+    let nanos_since_unix_epoch = nanos_since_unix_epoch.unwrap();
+
+    let duration = Duration::new(nanos_since_unix_epoch / 1_000_000_000,
+                                 (nanos_since_unix_epoch % 1_000_000_000) as u32);
+    Ok(UNIX_EPOCH + duration)
+}
+
+fn crate_path_tcx(tcx: TyCtxt, cnum: ast::CrateNum) -> PathBuf {
+    crate_path(tcx.sess, &tcx.crate_name(cnum), &tcx.crate_disambiguator(cnum))
+}
+
+/// Finds the correct metadata hashes file for the given crate. In order to do
+/// that it has to find the crate directory of the given crate, and in there,
+/// look for the session directory with the correct SVH in it.
+pub fn find_metadata_hashes_for(tcx: TyCtxt, cnum: ast::CrateNum) -> Option<PathBuf> {
+    let crate_directory = crate_path_tcx(tcx, cnum);
+
+    if !crate_directory.exists() {
+        return None
+    }
+
+    let dir_entries = match crate_directory.read_dir() {
+        Ok(dir_entries) => dir_entries,
+        Err(e) => {
+            tcx.sess
+               .err(&format!("incremental compilation: Could not read crate directory `{}`: {}",
+                             crate_directory.display(), e));
+            return None
+        }
+    };
+
+    let target_svh = tcx.sess.cstore.crate_hash(cnum).to_string();
+
+    let sub_dir = find_metadata_hashes_iter(&target_svh, dir_entries.filter_map(|e| {
+        e.ok().map(|e| e.file_name().to_string_lossy().into_owned())
+    }));
+
+    sub_dir.map(|sub_dir_name| crate_directory.join(&sub_dir_name)
+                                              .join(METADATA_HASHES_FILENAME))
+}
+
+fn find_metadata_hashes_iter<'a, I>(target_svh: &str, iter: I) -> Option<OsString>
+    where I: Iterator<Item=String>
+{
+    for sub_dir_name in iter {
+        if sub_dir_name.ends_with("-working") || !sub_dir_name.starts_with("sess-") {
```

Comment:
might be nice to encapsulate this into a helper, not sure how often it's repeated